### PR TITLE
[PrepareForEmission] Improve namehints heuristic and add mux heuristic

### PR DESCRIPTION
**Pull request by uenoku**
This PR improves namehints heuristic and add a heuristic to spill wires for mux. Namehint heuristic is modified to spill wires for namehints that doesn't have "\_" prefix regardless of term sizes. For expressions with "\_" prefix, if the term is greater than a threshold, the expression is spilled to a wire. Also this PR adds a heuristic option to spill all mux.

### CODE REVIEW

File: include/circt/Support/LoweringOptions.h [hunk 1]
```diff
@@ -129,12 +129,19 @@ struct LoweringOptions {
...
+    SpillNone = 0,                    // Default
+    SpillLargeTermsWithNamehints = 1, // Spill wires for expressions with
+                                      // namehints if the term size is greater
+                                      // than `wireSpillingNamehintTermLimit`.
+    SpillAllMux = 1 << 1,             //  Spill wires for all ternary mux.
+  };
+
+  unsigned wireSpillingHeuristicSet = SpillNone;
+
+  bool isWireSpillingHeuristicEnabled(WireSpillingHeuristic heurisic) const {
```
Comment (dtzSiFive) [inline include/circt/Support/LoweringOptions.h:142]: Should this have handling for `SpillNone` case?

File: lib/Support/LoweringOptions.cpp [hunk 2]
```diff
@@ -109,9 +110,10 @@ void LoweringOptions::parse(StringRef text, ErrorHandlerT errorHandler) {
       }
     } else if (option.consume_front("wireSpillingHeuristic=")) {
       if (auto heuristic = parseWireSpillingHeuristic(option)) {
-        wireSpillingHeuristic = *heuristic;
+        wireSpillingHeuristicSet |= *heuristic;
```
Comment (dtzSiFive) [inline lib/Support/LoweringOptions.cpp:113]: Error message and parsing logic suggest you can set `=spillNone`-- which doesn't have any effect (and maybe is strange if specified in conjunction with the other options, or at all, other than being explicitly about the preferred behavior regardless of defaults).
Not big deal, but maybe it should clear/override existing options or not  be allowed to be specified to keep these constructive?

File: test/Conversion/ExportVerilog/prepare-for-emission.mlir [hunk 3]
```diff
@@ -131,13 +131,45 @@ module attributes {circt.loweringOptions =
                   "wireSpillingHeuristic=spillLargeTermsWithNamehints,wireSpillingNamehintTermLimit=3"} {
   // CHECK-LABEL: namehints
   hw.module @namehints(%a: i8) -> (b: i8) {
-    // "foo" should not be spilled because the term size is 2.
-    // CHECK-NOT: %foo = sv.wire
+    // "foo" should be spilled because it has a meaningfull name.
```
Comment (dtzSiFive) [inline test/Conversion/ExportVerilog/prepare-for-emission.mlir:134]:
```suggestion
    // "foo" should be spilled because it has a meaningful name.
```

**Review by dtzSiFive — APPROVED** [review]
Generally LGTM!

Added some nits.

File: lib/Support/LoweringOptions.cpp (hunk 2 above)
Comment (uenoku) [reply lib/Support/LoweringOptions.cpp:113]: Good points, probably we can just remove `spillNone` to make it simple.

**Review by mikeurbach — APPROVED** [review]
I think I agree with @dtzSiFive's comments about spillNone, and otherwise this looks good. Thanks for putting this together!

**Pull request merged**